docs/proposals/volumes.md

```diff
@@ -0,0 +1,373 @@
+3.  distributed filesystems
+
+#### The hostPath corner case:
+
+The hostPath volume should only be used by 'trusted' users, and the permissions of paths exposed
+into containers via hostPath volumes should always be managed by the cluster operator.  If the
+Kubelet managed ownership or SELinux context for hostPath volumes, a user who could create a
+hostPath volume could affect changes in the state of arbitrary paths within the host's filesystem.
+This would be a huge security hole, so we will consider hostPath a corner case that the kubelet
+should never perform ownership or label management for.
+
+#### Volumes derived from empty dir
+
+Empty dir and volumes derived from it are the opposite of hostPath.  Since they are created by the
+system, Kubernetes must always ensure that the ownership and SELinux context (when relevant) are
+set correctly for the volume to be usable.
+
+#### Distributed file systems
+
+Whether the ownership or SELinux label of a volume for a distributed file system should be managed
+by Kubernetes depends on the cluster operator's policy and the volume type.  Additionally, not all
+distributed file systems support `chown`, `chmod`, and `chcon` operations uniformly.  Our API should
+make it simple to express whether a particular volume should have these concerns managed by
+Kubernetes.  For simplicity's sake, we will assume that if a volume requires ownership management,
+that it also requires relabeling and vice-versa.
+
+#### API requirements
+
+From the above, we know that ownership management must be applied:
+
+1.  To some volume types always
+2.  To some volume types never
+3.  To some volume types *sometimes*
+
+For volume types where the need for ownership/label management is not at the discretion of the user,
+there should be no need (or way) to specify whether ownership management is required in the API.
+Conversely there must be a way to specify this for volume types where it depends on the user's
+needs.
+
+### Determining correct ownership
+
+There are two components of ownership management:
+
+1.  When does Kubernetes need to manage the ownership of a volume?
+2.  How does Kubernetes determine the ownership of a volume?
+
+#### When to determine ownership
+
+Whether Kubernetes should manage the ownership of a volume for a distributed filesystem depends upon
+both the file system type and the cluster operator's policy.  For example:
+
+1.  Some organizations will manage ownership of volumes externally to the cluster
+2.  It is not possible to securely `chown` or `chmod` paths within some distributed filesystems
+
+#### How to determine ownership
+
+The Kubelet must analyze the pod spec to determine which UIDs need to use which volumes.  If a
+container's security context's `RunAsUser` field is not set, the Kubelet must inspect the image via
+the container runtime to determine which UID the image will run as.  Once the list of UIDs that need
```

Comment:
I think the kubelet should avoid inspecting container images. If `RunAsUser` is not specified then rely on the GID to grant volume permissions